designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/UIProcessValidator.scala

```diff
@@ -74,28 +74,49 @@ class UIProcessValidator(
       canonical: CanonicalProcess,
       processingType: ProcessingType
   )(implicit loggedUser: LoggedUser): ValidationResult = {
+    def validateAndFormatResult(scenario: CanonicalProcess) = {
+      val validated = validator.validate(scenario)
+      validated.result
+        .fold(formatErrors, _ => ValidationResult.success)
+        .withNodeResults(validated.typing.mapValuesNow(nodeInfoToResult))
+    }
+
     // TODO: should we validate after resolve?
     val additionalValidatorErrors = additionalValidators
       .map(_.validate(canonical))
       .sequence
       .fold(formatErrors, _ => ValidationResult.success)
 
-    val resolveResult = fragmentResolver.resolveFragments(canonical, processingType) match {
-      case Invalid(e) => formatErrors(e)
-      case _          =>
-        /* 1. We remove disabled nodes from canonical to not validate disabled nodes
-           2. TODO: handle types when fragment resolution fails... */
-        fragmentResolver.resolveFragments(canonical.withoutDisabledNodes, processingType) match {
-          case Valid(process) =>
-            val validated = validator.validate(process)
-            // FIXME: Validation errors for fragment nodes are not properly handled by FE
-            validated.result
-              .fold(formatErrors, _ => ValidationResult.success)
-              .withNodeResults(validated.typing.mapValuesNow(nodeInfoToResult))
-          case Invalid(e) => formatErrors(e)
+    val resolvedScenarioResult = fragmentResolver.resolveFragments(canonical, processingType)
+
+    // TODO: handle types when fragment resolution fails
+    val validationResult = resolvedScenarioResult match {
+      case Invalid(fragmentResolutionErrors) => formatErrors(fragmentResolutionErrors)
+      case Valid(scenario) =>
+        val validationResult = validateAndFormatResult(scenario)
+
+        val containsDisabledNodes =
+          canonical.withoutDisabledNodes.collectAllNodes.size != canonical.collectAllNodes.size
+        if (containsDisabledNodes) {
+
+          val resolvedScenarioWithoutDisabledNodes =
+            fragmentResolver.resolveFragments(canonical.withoutDisabledNodes, processingType)
+
+          resolvedScenarioWithoutDisabledNodes match {
+            case Invalid(e)                          => formatErrors(e)
+            case Valid(scenarioWithoutDisabledNodes) =>
+              // FIXME: Validation errors for fragment nodes are not properly handled by FE
+              // We add typing data from disabled nodes to have typing and suggestions for expressions in disabled nodes
+              val resultWithoutDisabledNodes = validateAndFormatResult(scenarioWithoutDisabledNodes)
+              resultWithoutDisabledNodes.copy(nodeResults =
+                resultWithoutDisabledNodes.nodeResults ++ validationResult.nodeResults
```

Comment:
```suggestion
                validationResult.nodeResults
```

resultWithoutDisabledNodes is (or should be) a subset of it anyway

Reply:
thanks for catching that. I'm pretty sure that there was some case where it was not the case, but cannot reproduce it now so it's probably just a mistake